### ISSUE TITLE
Custom ValueSource erroneously tracks files etc used in obtain() when ValueSource parameters are passed in

### ISSUE THREAD

**Issue by cloudshiftchris** [gradle/gradle#22305]
When creating a custom ValueSource with parameters, and using other ValueSource instances for those parameters, the inputs are erroneously tracked.

### Expected Behavior
As documented, ValueSource implementations should not track files, environment variables, etc.

### Current Behavior
Gradle erroneously tracks ValueSource inputs, resulting in unexpectedly dirty configuration caches.

### Context
This causes the configuration cache to miss 100% of the time.

  This is due to a collision of events:
     - Gradle fingerprints the input parameters to the ValueSource;
     - this results in beforeValueObtained() event being fired, which DISABLES input tracking
     - ...and then afterValueObtained() is fired, which ENABLES input tracking, just in time to drop into the obtain()
       method of our custom ValueSource

### Steps to Reproduce 
Reproduced [here](https://github.com/cloudshiftinc/repro-gradle-configuration-cache-value-source).  Consistently reproducible in configurations where a ValueSource is used as a parameter to a ValueSource

### Your Environment
MacOS, Gradle 7.5.1, Java 17

**Comment by cloudshiftchris** [comment]
At a minimum it should be documented that ValueSources should not be used as parameters to ValueSources; this is likely insufficient as it's not directly evident what constitutes a 'ValueSource' - it's intuitive to pass in providers.gradleProperty, for example.

A hard error when this is detected would be a good initial step.

Ideally the inputTracking state would be smart enough to support these use cases.

**Comment by cloudshiftchris** [comment]
Added additional diagnostic output to show the input tracking state:

```
        beforeValueObtained()  // input tracking turned off
        beforeValueObtained()  // input tracking turned off
        afterValueObtained()   // input tracking turned on
        valueObtained(): class org.gradle.api.internal.provider.sources.EnvironmentVariableValueSource org.gradle.api.internal.provider.sources.EnvironmentVariableValueSource$Inject@46d4f344

        // whoops; inside our ValueSource with input tracking turned on
        In value source
        Test file @ /Users/chrislee/.cloudshift/test.properties

        afterValueObtained()   // input tracking turned on
        valueObtained(): class Build_gradle$MyValueSource Build_gradle$MyValueSource$Inject@6383cdd1
```

**Comment by cloudshiftchris** [comment]
This looks to be fixable in org.gradle.configurationcache.fingerprint.ConfigurationCacheFingerprintWriter by changing the ThreadLocal<Boolean> to a counter (e.g. ThreadLocal<AtomicLong>) to support nested ValueSources, with a state of '0' -> input tracking enabled and < 0 (assuming beforeValueObtained() decrements counter) -> input tracking disabled.

**Comment by mlopatkin** [comment]
This will be fixed in 7.6 RC1. There is some follow-up work to do, tracked in #22494. Big thanks to @cloudshiftchris for finding the root cause!
